ships/0028-buildrun-cleanup.md

```diff
@@ -0,0 +1,175 @@
+
+Updates to TTL or limit fields in the Build after a BuildRun was created/completed will have the following behavior:
+
+- If a TTL field is updated, then no changes are made to BuildRuns. A TTL therefore applies to an individual BuildRun. This is consistent with timeouts that are also only considered at BuildRun creation.
+- If a limit field is updated, then this new value applies. The values in the BuildRun's `status.buildSpec.retention.*Limit` are therefore obsolete. A limit is applicable to all BuildRuns of a Build in a namespace. Handling this differently is not possible as there can only be a single limit for them.
+
+This means that TTL and limit behave slightly different which needs to be explained in the documentation.
+
+### Backend
+
+We will introduce new controllers for the BuildRun cleanup inside the same manager. The reasons for this are that our BuildRun controller is already highly complex - both in its predicate functions as well as in its `Reconcile` function - as it must handle TaskRuns and BuildRuns in various possible state combinations. There is one simple scenario where it simply does nothing: when the BuildRun is already completed. Instead of adding additional complexity for completed BuildRuns, the new controllers will focus exclusively on those aspects.
+
+We will introduce a new `buildrun-ttl-cleanup-controller` which will reconcile completed BuildRuns. It will watch BuildRuns with those predicate functions:
+
+- Create: if the BuildRun is completed (`status` of condition `type=Succeeded` set to `"true"` or `"false"`), and if a applicable TTL is set. This is handled for managers becoming the leader.
+- Update: if the BuildRun completed (`status` of condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if an applicable TTL is set.
+
+In the `Reconcile` function, we will check the BuildRun: When the `ttlAfterFailed` field is set for a BuildRun that failed, or when the `ttlAfterSucceeded` field is set for a BuildRun that succeeded, then we will calculate `.status.conditions[type=Succeeded].lastTransitionTime + TTL - now`. If the value is smaller than or equal zero, then we will delete the BuildRun. Otherwise, we will return a reconciliation result where `Requeue` is set to `true` and `RequeueAfter` is set to the calculated duration.
+
+We will introduce a new `build-limit-cleanup-controller` which will reconcile Builds. It will watch Builds with those predicate functions:
+
+- Create: if the Build has limits configured. This is handled for managers becoming the leader.
+- Update: if the Build has limits introduced, or changed to lower values.
+
+It will watch BuildRuns with those predicate functions:
+
+- Update: if the BuildRun completed (`status` of condition `type=Succeeded` is changed from `"unknown"` to `"true"` or `"false"`), and if it is related to a Build (standalone BuildRuns when introduced with SHIP 0029 will be ignored).
```

Comment:
+1 to this as well - standalone `BuildRun` objects can be considered "orphans" and should only use TTL for cleanup.